scipy/stats/_distn_infrastructure.py

```diff
@@ -538,20 +540,37 @@ def argsreduce(cond, *args):
     >>> C = rand((1, 5))
     >>> cond = np.ones(A.shape)
     >>> [A1, B1, C1] = argsreduce(cond, A, B, C)
+    >>> >>> A1.shape
+    (4, 5)
     >>> B1.shape
-    (20,)
+    (1,)
+    >>> C1.shape
+    (1, 5)
     >>> cond[2,:] = 0
-    >>> [A2, B2, C2] = argsreduce(cond, A, B, C)
-    >>> B2.shape
+    >>> [A1, B1, C1] = argsreduce(cond, A, B, C)
+    >>> A1.shape
+    (15,)
+    >>> B1.shape
+    (1,)
+    >>> C1.shape
     (15,)
-
     """
-    newargs = np.atleast_1d(*args)
+    newargs = np.atleast_1d(*args) # some distributions assume arguments are iterable.
     if not isinstance(newargs, list):
-        newargs = [newargs, ]
-    expand_arr = (cond == cond)
-    return [np.extract(cond, arr1 * expand_arr) for arr1 in newargs]
-
+        newargs = [newargs, ]           # np.atleast_1d returns an array if only one argument, or a list of arrays if more than one argument.
+    if np.all(cond):
+        return newargs         # Nothing to do
+
+    s = cond.shape
+    # np.extract returns flattened arrays, which are not broadcastable together unless they
+    # are either the same size or size == 1.
+    ret = []
+    for arg in newargs:
+        if np.size(arg) == 1:
+            ret.append(arg)
+        else:
+            ret.append(np.extract(cond, np.broadcast_to(arg, s)))
```

Comment:
I would keep the list comprehension (with a ternary here)
```suggestion
    ret = [(arg if np.size(arg) == 1
            else np.extract(cond, np.broadcast_to(arg, s)))
           for arg in newargs]
```